feat: adds google pay guide (#512)
Basis-Theory/developers.basistheory.com: docs/guides/google-pay/accept-payments.mdx

| [`PAN_ONLY`](https://developers.google.com/pay/api/web/guides/resources/payment-data-cryptography#pan_only)             | `card`            | The payment method stored is an FPAN (Funding Primary Account Number). This payment method is the underlying card itself and we have the capability to update the tokens tied to those payments with our [Account Updater](/docs/features/account-updater)                                                     |

Remember that Token Intents are short-lived and expire after 24 hours by default. If you're wanting to use the payment data stored in a Token Intent for recurring transactions, for example, you can convert the Token Intent to a [Token](/docs/concepts/what-are-tokens) by calling the [create Token endpoint](/docs/api/tokens/token-intents#convert-token-intent-to-token).

Let's try clicking the "GPay" button again and then the "Continue" button once again. You should see a Token Intent printed to your console now.

## Processing the Google Pay Payment Data

> We'll send the Basis Theory Token Intent to [Checkout](https://www.checkout.com/) to test charging a card. We'll be doing this from our backend by using a [Proxy](/docs/api/proxies/ephemeral-proxy) and a Private Application. Although, we are also able to create and authenticate a Basis Theory [Session](/docs/api/applications/sessions) and make this request from the frontend as well.

Reviewer: Lets make it generic and not mention Checkout at this point. 
Also, using sessions is not common and this will confuse customers.
```suggestion
We'll send the Basis Theory Token Intent to a Payments Service Provider (PSP) to test charging a card. We'll be doing this from our backend by using a [Proxy](/docs/api/proxies/ephemeral-proxy) and a Private Application. 
```

### Create a Private Application

> <ApplicationSection type="private" permissions={["token:use"]}/>

Reviewer: I thought this was going to be `token-intent:use`.

Author: not for lack of trying. unfortunately, that's not possible to implement at this time https://basistheory.slack.com/archives/C04R7KQQVUJ/p1739377636438479

        const paymentData = await paymentsClient.loadPaymentData(paymentDataRequest);
        console.log('Payment data:', paymentData);

        let encryptedPaymentToken = paymentData.paymentMethodData.tokenizationData.token;
        let tokenIntent = await tokenizeGooglePayTokenWithBasisTheory(encryptedPaymentToken);

        console.log(JSON.stringify(tokenIntent));
// highlight-start
        let { status } = await chargePayment(tokenIntent.id);

        if (status === 201) {
            console.log('Payment successful!');
        } else {
            console.error('Payment failed.');
        }
    } catch (error) {
        console.error('Error loading payment data:', error);
    }
}

// Charge Payment in backend
async function chargePayment(tokenIntentId) {
    const response = await fetch('http://localhost:3000/charge-payment', {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
        },
        body: JSON.stringify({
            tokenIntentId: tokenIntentId
        })
    });

    return await response.json();
}
// highlight-end
```

We'll setup and invoke the [ephemeral Proxy](/docs/api/proxies/ephemeral-proxy) on our backend.

> <Tabs className="bt-tabs" groupId="languages">

Reviewer: ```suggestion
<Tabs className="bt-tabs" groupId="processor">
```

-H 'Authorization: Bearer <CHECKOUT_AUTH_TOKEN>' \
// highlight-end
-d '{
  "source": {
    "type": "network_token",
    "token_type": "googlepay",
// highlight-start
    "token": "{{ token_intent: <TOKEN_INTENT_ID> | json: \"$.data.number\" }}",
    "expiry_month": "{{ token_intent: <TOKEN_INTENT_ID> | json: \"$.data\" | card_exp: \"MM\" }}",
    "expiry_year": "{{ token_intent: <TOKEN_INTENT_ID> | json: \"$.data\" | card_exp: \"YYYY\" }}"
// highlight-end
  },
  "amount": 1000,
  "currency": "USD",
  "processing_channel_id": "<CHECKOUT_PROCESSING_CHANNEL_ID>"
}'
```
  [Checkout Payment Method Docs](https://www.checkout.com/docs/payments/add-payment-methods/google-pay#Pay_with_a_pre-decrypted_Google_Pay_token).

You may notice that in this example we are not passing in any 3DS cryptogram or eci indicator. This is not necessary for charging FPANs. If you need to charge a DPAN, you will need to pass in the 3DS cryptogram and eci indicator available on the `authentication` property of the Token Intent.

```json showLineNumbers title="Create a payment transaction for a DPAN"
{
    "source": {
        "type": "network_token",
        "token_type": "googlepay",

>         "token": "{{ token_intent: <TOKEN_INTENT_ID> | json: \"$.data.number\" }}",

Reviewer: Highlight should start in this line

Author: I had it lower starting on `eci` because that's the part that's different from the highlighted section above. this part talks about adding `eci` and `cryptogram`

Reviewer: Oh yea, the DPAN vs FPAN thing. So smart.

                transactionInfo: {
                    totalPriceStatus: 'FINAL',
                    totalPrice: '10.00',
                    currencyCode: 'USD'
                },
                merchantInfo: {
                    merchantName: 'Example Merchant'
                }
            });

            try {
                const paymentData = await paymentsClient.loadPaymentData(paymentDataRequest);
                console.log('Payment data:', paymentData);
            } catch (error) {
                console.error('Error loading payment data:', error);
            }
        }
    </script>
</head>
<body>
<h1>Google Pay Integration</h1>
<div id="container"></div>
</body>
</html>
```

Note that for the `tokenizationSpecification`, we're setting `gateway` to `basistheory` and `gatewayMerchantId` to your `tenantId` created from the [Basis Theory Setup](#basis-theory-setup).

Now when we run this code, we should see a Google Pay button on our page. Clicking the "GPay" button will present a payment sheet. Click on the "Continue" button and take a look at the console to see the encrypted payment data.

## Storing the Google Pay Payment Data with Basis Theory

> Now that we have access to the encrypted payment data, we'll create a public key and send the payment data to Basis Theory using the [Google Pay connections endpoint](/docs/api/orchestration/google-pay) in the `onGooglePaymentButtonClicked` function.

Reviewer 2: ```suggestion
Now that we have access to the encrypted payment data, we'll create a Public Application and send the payment data to Basis Theory using the [Google Pay connections endpoint](/docs/api/orchestration/google-pay) in the `onGooglePaymentButtonClicked` function.
```

> ### Create a Public Key

Reviewer 2: ```suggestion
### Create a Public Application
```

        },
        merchantInfo: {
            merchantName: 'Example Merchant'
        }
    });

    try {
        const paymentData = await paymentsClient.loadPaymentData(paymentDataRequest);
        console.log('Payment data:', paymentData);

// highlight-start
        let encryptedPaymentToken = paymentData.paymentMethodData.tokenizationData.token;
        let tokenIntent = await tokenizeGooglePayTokenWithBasisTheory(encryptedPaymentToken);

        console.log(JSON.stringify(tokenIntent));
    } catch (error) {
        console.error('Error loading payment data:', error);
    }
}

// Tokenize Google Pay data using Basis Theory
async function tokenizeGooglePayTokenWithBasisTheory(encryptedPaymentToken) {
    const response = await fetch('https://api.basistheory.com/connections/google-pay/tokenize', {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
            'BT-API-KEY': '<PUBLIC_API_KEY>'
        },
        body: JSON.stringify({
            google_payment_method_token: encryptedPaymentToken
        })
    });

    let jsonResponse = await response.json();

    return jsonResponse.token_intent;
}
// highlight-end
```

> The result of this call will be a [Token Intent](/docs/api/tokens/token-intents#token-intent-object) that you can use to process the payment. The type of Token Intent will either be `card` or `network_token` depending on the `authMethod` of the Google Pay token that was available for the chosen card. The [`authMethod`](https://developers.google.com/pay/api/web/guides/resources/payment-data-cryptography#pan_only) is how the card was authenticated for the transaction. If the `authMethod` is of type [PAN_ONLY](https://developers.google.com/pay/api/web/guides/resources/payment-data-cryptography#pan_only), Google will send an encrypted FPAN (Funding Primary Account Number) and you will get a `card` type Token Intent from Basis Theory. If the `authMethod` of type [CRYPTOGRAM_3DS](https://developers.google.com/pay/api/web/guides/resources/payment-data-cryptography#cryptogram_3ds), Google will send an encrypted DPAN (Device Primary Account Number) and you will get a `network_token` Token Intent from Basis Theory. This is significant because if you store an encrypted DPAN from Google, you will get a payment method which is tied to the device for which it was added. This needs to be top of mind when charging with these payment methods in the future because we don't have insight into if these payment methods get revoked or are updated. If you store FPANs from Google with us, we have the capability to update the tokens tied to those payments with our [Account Updater](/docs/features/account-updater).

Reviewer 2: This might be summarized in a quick table with the Google Token Type -> Token Intent Type  -> description summary

Author: I'll chop this down

    response.json({
        result: chargeResponse.status,
    });
});

app.listen(3000, () => {
    console.log("Server is running on port 3000");
});
```

You may notice that in this example we are not passing in any 3DS cryptogram or eci indicator. This is not necessary for charging FPANs. If you need to charge a DPAN, you will need to pass in the 3DS cryptogram and eci indicator available on the `authentication` property of the Token Intent.

```json showLineNumbers title="Create a payment transaction for a DPAN"
{
  ...
  "mpiData": {
    "directoryResponse": "Y",
    "authenticationResponse": "Y",
// highlight-start
    "cavv": "{{ token_intent: d2cbc1b4-5c3a-45a3-9ee2-392a1c475ab4 | json: \"$.data.authentication.threeds_cryptogram\" }}",
    "eci": "{{ token_intent: d2cbc1b4-5c3a-45a3-9ee2-392a1c475ab4 | json: \"$.data.authentication.eciIndicator\" }}"
// highlight-end
  },
  "additionalData": {
    "paymentdatasource.type": "googlepay",
// highlight-start
    "paymentdatasource.tokenized": "true"
// highlight-end
  },
  ...
}
```

<Alert>You are not restricted to only using Adyen. You can use any payment processor that accepts Google Pay tokens.</Alert>

> ## Preparing for Production

Reviewer 2: I think we probably want to tell them how to convert the Token Intent into a real Token here and why they would do that.

"to store the FPAN or DPAN forever with Basis Theory...."

Author: I added a sentence about this. but I did it below the section where I talk about type of resource you get back from our GPay endpoint because it flowed better

2. Decryption of the payment data - The encrypted payment data is then sent to Basis Theory for decryption and storage of the sensitive cardholder information. Then Basis Theory returns a [Token Intent](/docs/api/tokens/token-intents#token-intent-object) that can be used to process the payment.
3. Processing of the payment data - The Basis Theory Token Intents are forwarded to the payment processor via Basis Theory Proxy, which translates Token Intents back to raw data before sending the request. When the successful payment response returns, the client application can complete the payment and inform the user that the payment was successful.

## Getting Started

The first steps will be to set up your Basis Theory account and ensure you can use your Google account for making transactions. Then we'll get into rendering the Google Pay button on your web page.

### Basis Theory Setup

<GettingStartedSection/>

> ### Google Pay Setup

Author: I think we can add Android down the road!

        let tokenIntent = await tokenizeGooglePayTokenWithBasisTheory(encryptedPaymentToken);

        console.log(JSON.stringify(tokenIntent));
// highlight-start
        let response = await chargePayment(tokenIntent.id);

        if (response.status === 200) {
            console.log('Payment successful!');
        } else {
            console.error('Payment failed.');
        }
    } catch (error) {
        console.error('Error loading payment data:', error);
    }
}

// Charge Payment in backend
async function chargePayment(tokenIntentId) {
    const response = await fetch('http://localhost:3000/charge-payment', {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
        },
        body: JSON.stringify({
            tokenIntentId: tokenIntentId
        })
    });

    return await response.json();
}
// highlight-end
```

In Adyen, you can charge a card by creating a [payment](https://docs.adyen.com/api-explorer/Checkout/latest/post/payments). We'll setup and invoke the [ephemeral Proxy](/docs/api/proxies/ephemeral-proxy) on our backend.

```javascript showlineNumbers title="Charge Payment method through a Proxy"

> const express = require("express");

Author: I wanted this thing to be the realest real boy. I had thoughts of publishing this possibly https://github.com/Basis-Theory-Labs/google-pay-web-example

> const express = require("express");

Reviewer: I appreciate that and think we should link that example in the guide.
But for support reasons,  it is best that we either go generic with `curl`, or we have examples in all "supported languages" for all the processors we will outline here.

    response.json({
        result: chargeResponse.status,
    });
});

app.listen(3000, () => {
    console.log("Server is running on port 3000");
});
```

You may notice that in this example we are not passing in any 3DS cryptogram or eci indicator. This is not necessary for charging FPANs. If you need to charge a DPAN, you will need to pass in the 3DS cryptogram and eci indicator available on the `authentication` property of the Token Intent.

```json showLineNumbers title="Create a payment transaction for a DPAN"
{
  ...
  "mpiData": {
    "directoryResponse": "Y",
    "authenticationResponse": "Y",
// highlight-start
    "cavv": "{{ token_intent: d2cbc1b4-5c3a-45a3-9ee2-392a1c475ab4 | json: \"$.data.authentication.threeds_cryptogram\" }}",
    "eci": "{{ token_intent: d2cbc1b4-5c3a-45a3-9ee2-392a1c475ab4 | json: \"$.data.authentication.eciIndicator\" }}"
// highlight-end
  },
  "additionalData": {
    "paymentdatasource.type": "googlepay",
// highlight-start
    "paymentdatasource.tokenized": "true"
// highlight-end
  },
  ...
}
```

<Alert>You are not restricted to only using Adyen. You can use any payment processor that accepts Google Pay tokens.</Alert>

> ## Preparing for Production

Reviewer: This should be an item in the Prod Checklist
